### DISCUSSION TITLE
Cannot understand how do I extract upload details and send it with request via params

### DISCUSSION THREAD

**Discussion by rohanm19** [rpldy/react-uploady#224] [Q&A]
Hi All,

I am trying to use Uploady for a react app, and I am failing to understand how do I extract details like file name and contents to send to my node server.

In the following snippet, QueueItem is where I extract the information. the console logs look fine. But since it is part of UploadPreview props, I am not sure how the parent prop "destination" can access this before rendering.

Not sure if I am able to clearly articulate the problem. But for me both the filename and body is not set, I just get a blank file with 0 as the filename

`
useItemProgressListener(item => {
      if (item.completed > progress) {
        setProgress(() => item.completed);
        setItemState(() =>
          item.completed === 100 ? STATES.DONE : STATES.PROGRESS
        );
      }
      console.log("item",item)
      setFileName(item.file.name);
      
      fetchAsBlob(props.url)
     .then(convertBlobToBase64)
     .then(setUploadBody)
     .then(console.log("uploadBody",uploadBody))
     .then(setIsLoading(false))

    }, props.id);
`

`<Uploady
          // debug
          destination={{ url: '/s3createObject', headers: {'filekey': nodePath[nodePath.length-1].Key + fileName, 
                                                            'accept': '*.jpg'},
                                                  params : {'body': uploadBody}} }
          // enhancer={composeEnhancers(mockSenderEnhancer)}
          >
            {/* <UploadProgress/> */}
          <UploadyButton></UploadyButton>
          <PreviewsContainer>
            <UploadPreview rememberPreviousBatches PreviewComponent={QueueItem} />
          </PreviewsContainer>
        </Uploady>`

**Comment by yoavniran** [comment]
hi @rohanm19 

Im not sure exactly what you're trying achieve. 

When uploading with Uploady, the default upload is done using a POST request with [formdata](https://developer.mozilla.org/en-US/docs/Web/API/FormData). 
The browser takes care of sending the file content and its info to the server:

<img width="783" alt="Screen Shot 2021-08-08 at 13 01 10" src="https://user-images.githubusercontent.com/1102278/128628229-1ef32c49-15e1-44e7-8371-afca0cb62c06.png">

You need to make sure your server knows how to handle multipart uploads (ex: [Multer](https://github.com/expressjs/multer) for Express on Node).

Any data you wish to add to the upload can be done using the _params_ prop. Dynamically adding params can be done easily with the 
[useRequestPreSend](https://github.com/rpldy/react-uploady/tree/master/packages/ui/uploady#userequestpresend-event-hook) hook.

Check out the [Dynamic Parameters Guide](https://github.com/rpldy/react-uploady/blob/master/guides/DynamicParameters.md) for sample code.

Let me know if something is still unclear

**Reply by x-N0** [reply]
@yoavniran I think what he wants to do its to send it as json body instaed of formdata (multipart form), is that possible with the latest version? 

How can you do something like this with useRequestPreSend?

**Reply by yoavniran** [reply]
Changing the format of the request cannot be done with useRequestPreSend. 
**A lower-level change is required**.

There's a (yet-to-be-documented) feature of the [xhr-sender](https://react-uploady.org/docs/packages/rpldy-sender/) that can be leveraged.
It's the ability to [create](https://github.com/rpldy/react-uploady/blob/master/packages/core/sender/types/index.d.ts#L33) an instance of the sender with a customized request body. 

Something like this:


```javascript
import { Uploady } from "@rpldy/uploady";
import { getXhrSend } from "@rpldy/sender";

const mySend = getXhrSend({
  getRequestData: (items, options) => {
                               // !! return the json object you want to send here !!
                        },
});

const MyApp = () => {
    return (
         <Uploady send={mySend}>
              <RestOfMyApp/>
         </Uploady>
    );
};

```

> Note that if you do this, its completely up to you to include everything you want in the data sent to the server. Uploady will not be involved.
